jumanji/environments/packing/job_shop/improvement/compute_makespan.py

```diff
@@ -0,0 +1,303 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# -----------------------------------------------------------------------------
+# This implementation is inspired by the paper:
+# "Deep Reinforcement Learning Guided Improvement Heuristic for Job Shop Scheduling"
+# by Cong Zhang, Zhiguang Cao, Wen Song, Yaoxin Wu, Jie Zhang
+# (https://arxiv.org/abs/2211.10936).
+#
+# The logic implemented here corresponds to the mechanism described in
+# Section 4.4: "Message Passing for Calculating Schedule".
+# Based on the adjacency matrix of the disjunctive graph, it allows to compute
+# earliest, latest start times and makespan for a given schedule.
+# For clarity, the notations used are the ones of the paper.
+# -----------------------------------------------------------------------------
+
+from typing import Tuple
+
+import chex
+import jax
+import jax.numpy as jnp
+import jraph
+
+
+def update_node_fn(
+    nodes: chex.Array,
+    _sent_attributes: chex.Array,
+    received_attributes: chex.Array,
+    _globals: chex.Array,
+) -> chex.Array:
+    """Update the date and completion status of a node based on the received attributes.
+
+    Args:
+        nodes: Array of nodes with shape (num_nodes, 3)
+        _sent_attributes: Array of sent attributes with shape (num_edges, 2)
+        received_attributes: Array of received attributes with shape (num_edges, 2)
+        _globals: Array of globals with shape (num_globals, 1)
+    Returns:
+        Array of updated nodes with shape (num_nodes, 3)
+    """
+
+    # Each node has a feature (p, d, c) with p the processing time,
+    # d the date and c the completion status
+    node_p = nodes[:, 0]  # remains unchanged
+
+    # Update date and completion status if neighbours
+    received_d = received_attributes[:, 0]
+    received_c = received_attributes[:, 1]
+
+    return jnp.stack([node_p, received_d, received_c], axis=-1)
+
+
+def forward_pass_jraph(
```

Comment:
Since this computes the makespan, does it make sense to put it in the name ? 
It would help to understand why and where you should use this function